static/app/components/hookOrDefault.tsx

```diff
@@ -46,54 +46,58 @@ function HookOrDefault<H extends HookName>({
   hookName,
   defaultComponent,
   defaultComponentPromise,
-}: Params<H>) {
-  type Props = React.ComponentProps<ReturnType<Hooks[H]>>;
-  type State = {hooks: Hooks[H][]};
-
-  class HookOrDefaultComponent extends Component<Props, State> {
-    static displayName = `HookOrDefaultComponent(${hookName})`;
+}: Params<H>): React.FunctionComponent<ComponentProps<ReturnType<Hooks[H]>>> {
+  type Props = ComponentProps<ReturnType<Hooks[H]>>;
 
-    state: State = {
-      hooks: HookStore.get(hookName),
-    };
+  // Defining the props here is unnecessary and slow for typescript
+  function getDefaultComponent(): React.ComponentType<any> | undefined {
+    // If `defaultComponentPromise` is passed, then return a Suspended component
+    if (defaultComponentPromise) {
+      // Lazy adds a complicated type that is not important
+      const DefaultComponent: React.ComponentType<any> = lazy(defaultComponentPromise);
 
-    componentWillUnmount() {
-      this.unlistener?.();
+      return function (props: Props) {
+        return (
+          <Suspense fallback={null}>
+            <DefaultComponent {...props} />
+          </Suspense>
+        );
+      };
     }
 
-    unlistener = HookStore.listen(
-      (name: string, hooks: Hooks[HookName][]) =>
-        name === hookName && this.setState({hooks}),
-      undefined
-    );
+    return defaultComponent;
+  }
 
-    get defaultComponent() {
-      // If `defaultComponentPromise` is passed, then return a Suspended component
-      if (defaultComponentPromise) {
-        const DefaultComponent = lazy(defaultComponentPromise);
+  function HookOrDefaultComponent(props: Props) {
+    const [hooks, setHooks] = useState<Hooks[H][]>(HookStore.get(hookName));
 
-        return function (props: Props) {
-          return (
-            <Suspense fallback={null}>
-              <DefaultComponent {...props} />
-            </Suspense>
-          );
-        };
-      }
+    useEffect(() => {
+      const unlistener = HookStore.listen((name: string, newHooks: Hooks[H][]) => {
+        if (name === hookName) {
+          setHooks(newHooks);
+        }
+      }, undefined);
 
-      return defaultComponent;
-    }
+      return () => {
+        unlistener();
```

Comment:
nit: but naming is hard

```suggestion
        unsubscribe();
```